### ISSUE TITLE
TypeError: sequence item 0: expected str instance, NoneType found

### ISSUE THREAD

**Issue by felipe3dfx** [jazzband/django-axes#437]
When complete process with [social-app-django](https://github.com/python-social-auth/social-app-django), raise the following exception:

```python
TypeError: sequence item 0: expected str instance, NoneType found
  File "django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "django/views/decorators/cache.py", line 44, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "social_django/utils.py", line 49, in wrapper
    return func(request, backend, *args, **kwargs)
  File "social_django/views.py", line 33, in complete
    *args, **kwargs)
  File "social_core/actions.py", line 41, in do_complete
    user = backend.complete(user=user, *args, **kwargs)
  File "social_core/backends/base.py", line 40, in complete
    return self.auth_complete(*args, **kwargs)
  File "social_core/utils.py", line 252, in wrapper
    return func(*args, **kwargs)
  File "social_core/backends/oauth.py", line 405, in auth_complete
    *args, **kwargs)
  File "social_core/utils.py", line 252, in wrapper
    return func(*args, **kwargs)
  File "social_core/backends/oauth.py", line 416, in do_auth
    return self.strategy.authenticate(*args, **kwargs)
  File "social_django/strategy.py", line 107, in authenticate
    return authenticate(*args, **kwargs)
  File "django/contrib/auth/__init__.py", line 73, in authenticate
    user = backend.authenticate(request, **credentials)
  File "axes/helpers.py", line 393, in inner
    return func(*args, **kwargs)
  File "axes/backends.py", line 38, in authenticate
    if AxesProxyHandler.is_allowed(request, credentials):
  File "axes/handlers/proxy.py", line 45, in is_allowed
    return cls.get_implementation().is_allowed(request, credentials)
  File "axes/handlers/base.py", line 43, in is_allowed
    if self.is_locked(request, credentials):
  File "axes/handlers/base.py", line 102, in is_locked
    return self.get_failures(request, credentials) >= settings.AXES_FAILURE_LIMIT
  File "axes/handlers/cache.py", line 30, in get_failures
    cache_key = get_client_cache_key(request, credentials)
  File "axes/helpers.py", line 373, in get_client_cache_key
    cache_key_components = ''.join(filter_kwargs.values())
```

```python
AUTHENTICATION_BACKENDS = [
    'axes.backends.AxesBackend',
    'social_core.backends.facebook.FacebookOAuth2',
    'social_core.backends.google.GoogleOAuth2',
    'social_core.backends.linkedin.LinkedinOAuth',
    'django.contrib.auth.backends.AllowAllUsersModelBackend',
    'app.backends.IdentificationOrEmailBackend',
]
```

**Comment by aleksihakli** [comment]
This is most probably due to IP address resolution failing and returning a None as IP from the OAuth backend authentication. Does this result from logging in with Facebook, Google or LinkedIn?

**Comment by felipe3dfx** [comment]
Failed with Facebook and Google

**Comment by felipe3dfx** [comment]
@aleksihakli with Django authenticate will raise same error, i guess forgot something:

```python
TypeError: sequence item 0: expected str instance, NoneType found
  File "django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "django/utils/decorators.py", line 45, in _wrapper
    return bound_method(*args, **kwargs)
  File "django/views/decorators/debug.py", line 76, in sensitive_post_parameters_wrapper
    return view(request, *args, **kwargs)
  File "django/utils/decorators.py", line 45, in _wrapper
    return bound_method(*args, **kwargs)
  File "django/utils/decorators.py", line 142, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "django/utils/decorators.py", line 45, in _wrapper
    return bound_method(*args, **kwargs)
  File "django/views/decorators/cache.py", line 44, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "django/contrib/auth/views.py", line 61, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "django/views/generic/base.py", line 97, in dispatch
    return handler(request, *args, **kwargs)
  File "django/views/generic/edit.py", line 141, in post
    if form.is_valid():
  File "django/forms/forms.py", line 185, in is_valid
    return self.is_bound and not self.errors
  File "django/forms/forms.py", line 180, in errors
    self.full_clean()
  File "django/forms/forms.py", line 382, in full_clean
    self._clean_form()
  File "django/forms/forms.py", line 409, in _clean_form
    cleaned_data = self.clean()
  File "django/contrib/auth/forms.py", line 196, in clean
    self.user_cache = authenticate(self.request, username=username, password=password)
  File "django/contrib/auth/__init__.py", line 73, in authenticate
    user = backend.authenticate(request, **credentials)
  File "axes/helpers.py", line 393, in inner
    return func(*args, **kwargs)
  File "axes/backends.py", line 38, in authenticate
    if AxesProxyHandler.is_allowed(request, credentials):
  File "axes/handlers/proxy.py", line 45, in is_allowed
    return cls.get_implementation().is_allowed(request, credentials)
  File "axes/handlers/base.py", line 43, in is_allowed
    if self.is_locked(request, credentials):
  File "axes/handlers/base.py", line 102, in is_locked
    return self.get_failures(request, credentials) >= settings.AXES_FAILURE_LIMIT
  File "axes/handlers/cache.py", line 30, in get_failures
    cache_key = get_client_cache_key(request, credentials)
  File "axes/helpers.py", line 373, in get_client_cache_key
    cache_key_components = ''.join(filter_kwargs.values())
```

Configuration:
```python
MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',

    'app.middleware.CampaignMiddleware',

    'axes.middleware.AxesMiddleware',
    'debug_toolbar.middleware.DebugToolbarMiddleware',
]

AXES_FAILURE_LIMIT = 10
AXES_USE_USER_AGENT = True
AXES_COOLOFF_TIME = 24
AXES_LOCKOUT_TEMPLATE = '429.html'
AXES_CACHE = 'axes'
AXES_HANDLER = 'axes.handlers.cache.AxesCacheHandler'

CACHES = {
    'default': {
        'BACKEND': 'django_redis.cache.RedisCache',
        'LOCATION': f'redis://{REDIS_HOST}:{REDIS_PORT}/1',
        'OPTIONS': {
            'CLIENT_CLASS': 'django_redis.client.DefaultClient',
        }
    },
    'sessions': {
        'BACKEND': 'django_redis.cache.RedisCache',
        'LOCATION': f'redis://{REDIS_HOST}:{REDIS_PORT}/2',
        'OPTIONS': {
            'CLIENT_CLASS': 'django_redis.client.DefaultClient',
        }
    },
    'compress': {
        'BACKEND': 'django_redis.cache.RedisCache',
        'LOCATION': f'redis://{REDIS_HOST}:{REDIS_PORT}/3',
        'OPTIONS': {
            'CLIENT_CLASS': 'django_redis.client.DefaultClient',
        }
    },
    'axes': {
        'BACKEND': 'django_redis.cache.RedisCache',
        'LOCATION': f'redis://{REDIS_HOST}:{REDIS_PORT}/4',
        'OPTIONS': {
            'CLIENT_CLASS': 'django_redis.client.DefaultClient',
        }
    }
}
```

**Comment by aleksihakli** [comment]
@felipe3dfx this should be fixed in 30184e2e524f52f017d0e9e5b6d66f274a4dbc80, I added a regression test for the possibly `None` IP address in cache key calculation which was the source of the error. Release has been rolled out in version 5.0.4.

**Comment by felipe3dfx** [comment]
@aleksihakli Now works fine! thank you!
